pallets/validators-manager/src/lib.rs

```diff
@@ -640,26 +565,237 @@ impl<T: Config> Pallet<T> {
+        TotalIngresses::<T>::put(ingress_counter);
+
+        <ValidatorActions<T>>::insert(
+            validator_account_id,
+            ingress_counter,
+            ValidatorsActionData::new(
+                ValidatorsActionStatus::AwaitingConfirmation,
+                tx_id,
+                ValidatorsActionType::Resignation,
+            ),
+        );
+
+        Self::deposit_event(Event::<T>::PublishingValidatorActionOnEthereumSucceeded { tx_id });
+
+        Ok(tx_id)
+    }
 }
 
 impl<T: Config> BridgeInterfaceNotification for Pallet<T> {
     fn process_result(tx_id: u32, caller_id: Vec<u8>, succeeded: bool) -> DispatchResult {
-        // TODO: Update data structure to use tx_id as key
-        if caller_id == PALLET_ID.to_vec() {
-            if succeeded {
-                log::info!(
-                    "✅  Transaction with ID {} was successfully published to Ethereum.",
-                    tx_id
-                );
-                Self::deposit_event(Event::<T>::PublishingValidatorActionOnEthereumSucceeded {
-                    tx_id,
-                });
-            } else {
-                log::error!("❌ Transaction with ID {} failed to publish to Ethereum.", tx_id);
-                Self::deposit_event(Event::<T>::PublishingValidatorActionOnEthereumFailed {
-                    tx_id,
-                });
+        if caller_id != PALLET_ID.to_vec() {
+            return Ok(())
+        }
+
+        // Find the ValidatorActions entry with matching tx_id
+        let mut found_entry: Option<(T::AccountId, IngressCounter, ValidatorsActionType)> = None;
+        
+        for (account_id, ingress_counter, validators_action_data) in <ValidatorActions<T>>::iter() {
+            if validators_action_data.eth_transaction_id == tx_id {
+                found_entry = Some((
+                    account_id,
+                    ingress_counter,
+                    validators_action_data.action_type,
+                ));
+                break;
+            }
+        }
+
+        let Some((account_id, ingress_counter, action_type)) = found_entry else {
+            // No matching entry found - might have been cleaned up already
+            return Ok(())
+        };
+
+        if !succeeded {
+            // T1 operation failed - cleanup the action and rollback state
+            if action_type.is_registration() {
+                // Remove the eth key mapping we added optimistically
+                if let Some(eth_key) = Self::get_ethereum_public_key_if_exists(&account_id) {
+                    <EthereumPublicKeys<T>>::remove(eth_key);
+                }
+                // Cleanup stored deposit
+                PendingRegistrationDeposits::<T>::remove(&account_id);
             }
+            
+            <ValidatorActions<T>>::remove(&account_id, ingress_counter);
+            Self::deposit_event(Event::<T>::PublishingValidatorActionOnEthereumFailed { tx_id });
```

Comment:
The way its setup, you can end up with `PublishingValidatorActionOnEthereumSucceeded` and `PublishingValidatorActionOnEthereumFailed` for the same action. For example if the registration has been accepted and added to the queue on EthBridge we get the success event. Then if the transaction fails to execute we get the failed event. I think `PublishingValidatorActionOnEthereumSucceeded` should be replaced with: 
 - `ValidatorActionPublished`: sent to ethereum via eth bridge
 - `FailedToPublishValidatorAction`: failed to send to ethereum
 - `ValidatorActionConfirmedOnEthereum`: tx succeeded on ethereum
 - `ValidatorActionFailedOnEthereum`: tx failed on ethereum